apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex

```diff
@@ -0,0 +1,89 @@
+defmodule Lexical.RemoteControl.CodeMod.Rename.Prepare do
+  alias Lexical.Ast
+  alias Lexical.Ast.Analysis
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Lexical.RemoteControl.CodeIntelligence.Entity
+  alias Lexical.RemoteControl.CodeMod.Rename
+  alias Lexical.RemoteControl.Search.Store
+
+  require Logger
+
+  @spec prepare(Analysis.t(), Position.t()) ::
+          {:ok, {atom(), String.t()}, Range.t()} | {:error, term()}
+  def prepare(%Analysis{} = analysis, %Position{} = position) do
+    case resolve(analysis, position) do
+      {:ok, {:module, module}, range} ->
+        {:ok, inspect(module), range}
+
+      {:ok, {:call, {_m, f}}, range} ->
+        {:ok, to_string(f), range}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  @spec resolve(Analysis.t(), Position.t()) ::
+          {:ok, {atom(), atom()} | {atom(), tuple()}, Range.t()} | {:error, term()}
+  def resolve(%Analysis{} = analysis, %Position{} = position) do
+    case do_resolve(analysis, position) do
+      {:ok, {:module, _module}, _range} ->
+        {module, range} = surround_the_whole_module(analysis, position)
+
+        if rename_at_declaration?(module, range) do
+          {:ok, {:module, module}, range}
+        else
+          {:error, {:unsupported_location, :module}}
+        end
+
+      other ->
+        other
+    end
+  end
+
+  defp surround_the_whole_module(analysis, position) do
+    # When renaming occurs, we want users to be able to choose any place in the defining module,
+    # not just the last local module, like: `defmodule |Foo.Bar do` also works.
+    {:ok, %{end: {_end_line, end_character}}} = Ast.surround_context(analysis, position)
+    end_position = %{position | character: end_character - 1}
+    {:ok, {:module, module}, range} = do_resolve(analysis, end_position)
+    {module, range}
+  end
+
+  defp rename_at_declaration?(module, rename_range) do
+    case Store.exact(module, type: :module, subtype: :definition) do
+      [definition] ->
+        rename_range == definition.range
+
+      _ ->
+        false
+    end
+  end
+
+  @renamable_module [Rename.Module]
+
+  defp do_resolve(%Analysis{} = analysis, %Position{} = position) do
+    result =
+      Enum.find_value(@renamable_module, fn module ->
+        result = module.resolve(analysis, position)
+
+        if match?({:ok, _, _}, result) do
+          result
+        end
+      end)
+
+    if is_nil(result) do
+      case Entity.resolve(analysis, position) do
+        {:ok, other, _} ->
+          Logger.info("Unsupported entity for renaming: #{inspect(other)}")
+          {:error, :unsupported_entity}
+
+        {:error, reason} ->
+          {:error, reason}
+      end
```

Comment:
Yes, it's just for the *error message*, because I found that our `Entity.resolve` still has some missing cases. I'm inclined to keep this logic for a while.

Reply:
Doesn't `Entity.resolve` already have a log message if it can't resolve something? Why have two?

Reply:
Sorry, perhaps I didn't explain clearly enough.

The reason we need to call it twice here is that the *specific rename module is only concerned with the resolve results of that module*. For instance, the `module` module is only concerned with `module and struct` cases, while the `callable` module only cares about the `:call` situation.

So, when we use this code here, it only returns the entities we supported or `nil`. 

https://github.com/lexical-lsp/lexical/blob/8dc3c2f850ec1b266938b7c72cbcd9789aef3891/apps/remote_control/lib/lexical/remote_control/code_mod/rename/prepare.ex#L65-L72

The purpose of the second `resolve` is to identify which entities we do not support, something we can't see within the specific modules.

FYI: another callable rename branch:

https://github.com/lexical-lsp/lexical/blob/ee4dc9f909d308e13e128974539792ef2f8c489d/apps/remote_control/lib/lexical/remote_control/code_mod/rename/callable.ex#L15-L23